File: integration/fsharp/integration.fs
module integration = 
  type Fx = double -> double

  let integrate( a : double)(b : double)(g : int)(f : Fx) : double = 
    let interval = (b - a) / (double) g
    let fxValues = seq { for n in 1 .. g-1 -> f(a + (double)n * interval) }
    let fxSum = Seq.fold (+) 0.0 fxValues
    (b - a) / (double) g * (f(a) / 2.0 + f(b) / 2.0 + fxSum)

module test1 = 

  let xSquared x : double = x * x
  open integration
  let area = integrate 0.0 10.0 1000 xSquared
  printfn "Area = %f" area

module test = 

  let xSquared x : double = x * x
  open integration
  let area = integrate 0.0 100.0 1000 xSquared
  printfn "Area = %f" area

File: integration/scala/integration.scala
object integration extends App {

	type Fx = (Double) => Double

	def xSquared (x : Double) : Double = {
  		return x * x 
	}

	def integrate ( a : Double, b : Double, g : Int, f : Fx) : Double = {
		def FxValues(): List[Double] = {
			var interval = (b-a) / g;
    		for { n <- List.range(1 , g) } 
    		   	yield { f(a + n * interval) }
		}
		var fxSum = FxValues().foldLeft(0.0)(_ + _)
    	return (b - a) / g * (f(a) / 2.0 + f(b) / 2.0 + fxSum)
	}


	var area = integrate( 0, 10, 1000, xSquared)
	println("Area = " + area)
}


